---
title: "MetaMarkers: robust marker signatures from single-cell data (annotation vignette)"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MetaMarkers: robust marker signatures from single-cell data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center",
  fig.width = 6,
  fig.height = 4.5
)
```

## Introduction

We assume that the reader is familiar with the main MetaMarker vignette, which shows how to select meta-markers from a compendium of pancreas dataset. In this vignette, we will focus on using markers for cell type annotation, which also serves as a benchmark to find highly informative markers.

As in the main vignette, the requirements are:
 - a list of annotated datasets.
 - matching cell type names across datasets.
 - all packages installed in the main vignette.

## Prepare datasets

We load the packages needed for our analysis: MetaMarkers, scRNAseq (package used to access datasets) and dplyr (data manipulation, part of the tidyverse).

```{r setup, message=FALSE}
devtools::load_all("..")
library(scRNAseq)
library(dplyr)
```

We will perform our annotation task on pancreatic cell types. For simplicity, we only consider 3 datasets here, but the analysis can easily be extended to include even more datasets. We use leave-one-dataset-out cross-validation: we build meta-markers from 2 datasets and predict on the third dataset.

First, we load the datasets (in the SingleCellExperiment format) and make sure that gene names are identical across datasets. Here, the Baron dataset uses gene symbols, while the Muraro dataset uses names that combine gene symbols and chromosome locations (to avoid duplicate names). We change gene names to gene symbols, which are contained in the "rowData" slot of the Muraro dataset. Note that MetaMarkers will automatically remove duplicate gene names, so we don’t need to worry.

```{r load_datasets, message=FALSE}
dataset1 = scRNAseq::BaronPancreasData()
dataset2 = scRNAseq::MuraroPancreasData()
dataset3 = scRNAseq::SegerstolpePancreasData()
rownames(dataset2) = rowData(dataset2)$symbol
```

Next we match cell type names across datasets.

```{r display_cell_types}
table(dataset1$label)
table(dataset2$label)
table(dataset3$`cell type`)
```

Some cell types are only present in one dataset, but there is a good overlap for the main cell types. We update names of cell types that overlap but have different names.

```{r rename_cell_types}
new_label = dataset1$label
new_label[new_label=="activated_stellate"] = "mesenchymal"
new_label[new_label=="quiescent_stellate"] = "mesenchymal"
dataset1$new_label = new_label

new_label = dataset2$label
new_label[new_label=="duct"] = "ductal"
new_label[new_label=="pp"] = "gamma"
dataset2$new_label = new_label

new_label = dataset3$`cell type`
new_label = gsub(" cell", "", new_label)
new_label[new_label=="PSC"] = "mesenchymal"
dataset3$new_label = new_label

common_labels = Reduce(intersect, list(
    unique(dataset1$new_label), unique(dataset2$new_label), unique(dataset3$new_label)
))
common_labels
```

We only keep cell types that are common to both datasets. Note that this step is not strictly necessary: if we kept cell types present in only one dataset, the pipeline would run normally but the "meta-markers" for these cell types would be based on a single dataset.

```{r susbet_datasets}
dataset1 = dataset1[, dataset1$new_label %in% common_labels]
dataset2 = dataset2[, dataset2$new_label %in% common_labels]
dataset3 = dataset3[, dataset3$new_label %in% common_labels]
```
The last step before computing markers is data normalization. The only requirement here is that the normalization is uniform across datasets for the meta-analytic stats to be meaningful. Any normalization procedure may be used, the MetaMarkers package offer a simple library size normalization procedure:

```{r compute_cpm}
cpm(dataset1) = convert_to_cpm(assay(dataset1))
cpm(dataset2) = convert_to_cpm(assay(dataset2))
cpm(dataset3) = convert_to_cpm(assay(dataset3))
```

# Marker selection

We proceed to marker selection for each inidividual dataset

```{r compute_markers}
markers = list(
     baron = compute_markers(cpm(dataset1), dataset1$new_label),
     muraro = compute_markers(cpm(dataset2), dataset2$new_label),
     seger = compute_markers(cpm(dataset3), dataset3$new_label)
)
```

Then build meta-markers from invidual datasets marker lists, leaving one dataset out.

```{r compute_meta_markers}
meta_markers = lapply(names(markers), function(leave_out) {
    make_meta_markers(markers[names(markers) != leave_out])
})
names(meta_markers) = names(markers)
```

# Example of cell type prediction

As an example, we start by predicting cell types from the Segerstolpe dataset using the top 100 Baron+Muraro meta-markers. First we select the top 100 meta-markers (leaving Segerstolpe out) for each cell type:

```{r extract_top_markers}
top_markers = filter(meta_markers[["seger"]], rank<=100)
```

Then predict cell types in the Segerstolpe dataset:

```{r score_seger}
ct_scores = score_cells(log1p(cpm(dataset3)), top_markers)
ct_scores[, 1:3]
```

Each cell in the Segerstople dataset receives a score for each cell types. For example, cell "HP1502401_H13" scores highest for the gamma cell type.

To evaluate cell typing performance, we can use the Segerstolpe annotation as a ground truth. We can ask two separate questions:
 - how good are markers at separating populations?
 - how easily can we find a threshold to predict the correct cell type?

For the first question, we don’t need to predict cell types. Instead, for each cell type, we ask how far apart...

```{r auroc_seger}
true_labels = dataset3$new_label
auroc_seger = summarize_auroc(ct_scores, true_labels)
fc_seger = summarize_fold_change(ct_scores, true_labels)
gplots::heatmap.2(auroc_seger)
gplots::heatmap.2(log2(fc_seger))
```

For the second question, we need to convert scores to cell type predictions. Here we will simply assign cell to the highest marker score.

```{r predict_seger}
ct_pred = assign_cells(ct_scores)
head(ct_pred)
```

For example, the first cell has been predicted to be a gamma cell, "score" and "quality" are used to quantify the confidence of the prediction. NANAN alpha

```{r}
summarize_precision_recall(predictor = ct_pred == "alpha", is_positive = true_labels == "alpha", threshold=0)
```

```{r}
f1 = lapply(unique(true_labels), function(cell_type) {
    result = summarize_precision_recall(predictor = ct_pred == cell_type,
                                        is_positive = true_labels == cell_type,
                                        threshold=0)
    result$value[result$stat == "f1"]
})
names(f1) = unique(true_labels)
f1
```
Works extremely well, except acinar and epsilon.

```{r}
data.frame(epsilon_scores = ct_scores["epsilon",], true_label = true_labels) %>%
    ggplot(aes(x = epsilon_scores, fill = true_label)) +
    geom_density(alpha=0.4)

ct_scores[,true_labels == "epsilon"] %>%
    t() %>%
    as.data.frame() %>%
    pivot_longer(everything(),"prediction", values_to = "score") %>%
    ggplot(aes(x = score, fill = prediction)) +
    geom_density(alpha=0.4)
```

```{r}
epsilon_score = ct_scores["epsilon",] / colSums(ct_scores) 
summary = summarize_precision_recall(
    predictor = epsilon_score,
    is_positive = true_labels == "epsilon",
    threshold=seq(0,0.5,by=0.01)
)
summary %>%
    filter(stat == "f1") %>%
    ggplot(aes(x = threshold, y=value)) +
    geom_line()
```

Optimal F1 0.8, ...

# Estimation of number of informative markers per cell type